community/kernel/src/main/java/org/neo4j/kernel/impl/store/PropertyStore.java

```diff
@@ -226,91 +229,155 @@ public void encodeValue( PropertyBlock block, int keyId, Value value )
     public static void encodeValue( PropertyBlock block, int keyId, Value value,
             DynamicRecordAllocator stringAllocator, DynamicRecordAllocator arrayAllocator )
     {
-        // TODO: use ValueWriter
-        Object asObject = value.asObject();
-        if ( asObject instanceof String )
-        {   // Try short string first, i.e. inlined in the property block
-            String string = (String) asObject;
-            if ( LongerShortString.encode( keyId, string, block, PropertyType.getPayloadSize() ) )
+        if ( value instanceof ArrayValue )
+        {
+            Object asObject = value.asObject();
+
+            // Try short array first, i.e. inlined in the property block
+            if ( ShortArray.encode( keyId, asObject, block, PropertyType.getPayloadSize() ) )
             {
                 return;
             }
 
-            // Fall back to dynamic string store
-            byte[] encodedString = encodeString( string );
-            List<DynamicRecord> valueRecords = new ArrayList<>();
-            allocateStringRecords( valueRecords, encodedString, stringAllocator );
-            setSingleBlockValue( block, keyId, PropertyType.STRING, Iterables.first( valueRecords ).getId() );
-            for ( DynamicRecord valueRecord : valueRecords )
+            // Fall back to dynamic array store
+            List<DynamicRecord> arrayRecords = new ArrayList<>();
+            allocateArrayRecords( arrayRecords, asObject, arrayAllocator );
+            setSingleBlockValue( block, keyId, PropertyType.ARRAY, Iterables.first( arrayRecords ).getId() );
+            for ( DynamicRecord valueRecord : arrayRecords )
             {
-                valueRecord.setType( PropertyType.STRING.intValue() );
+                valueRecord.setType( PropertyType.ARRAY.intValue() );
             }
-            block.setValueRecords( valueRecords );
+            block.setValueRecords( arrayRecords );
         }
-        else if ( asObject instanceof Integer )
+        else
         {
-            setSingleBlockValue( block, keyId, PropertyType.INT, ((Integer) asObject).longValue() );
+            value.writeTo( new PropertyBlockValueWriter( block, keyId, stringAllocator ) );
         }
-        else if ( asObject instanceof Boolean )
+    }
+
+    private static class PropertyBlockValueWriter implements ValueWriter<IllegalArgumentException>
```

Comment:
we tried to use property block value writer here before but benchmarks where showing degradation because of that please double check that this can be done now

Reply:
Yes. No regressions from what we can see. An image is attached to the trello card.